### PR TITLE
[Core] Introduce Bill of Materials (BOM)

### PR DESCRIPTION
**Pull request by stepio**
Implemented Bill of Materials for https://github.com/cucumber/cucumber-jvm/issues/2044

Versions are inherited from the Cucumber's parent pom. Started with all the dependencies from the parent pom, added few additional ones from the other modules, rearranged them alphabetically by `artifactId`. Also added trivial README file to explain the module's usage.

For testing added the BOM into my pet project:
https://github.com/stepio/coffee-boots-example/blob/poc-cucumber/pom.xml
```xml
<dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>io.cucumber</groupId>
            <artifactId>cucumber-bom</artifactId>
            <version>6.10.1-SNAPSHOT</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>

<dependencies>
   <dependency>
        <groupId>io.cucumber</groupId>
        <artifactId>cucumber-core</artifactId>
        <scope>test</scope>
    </dependency>
</dependencies>
```
Then checked dependencies with the next command:
```sh
mvn dependency:tree
```
Got expected output:
```sh
[INFO] Scanning for projects...
[INFO] 
[INFO] -----------< io.github.stepio.examples:coffee-boots-example >-----------
[INFO] Building coffee-boots-example 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ coffee-boots-example ---
[INFO] io.github.stepio.examples:coffee-boots-example:jar:0.0.1-SNAPSHOT
[INFO] +- io.github.stepio.coffee-boots:coffee-boots:jar:2.2.1-SNAPSHOT:compile
[INFO] |  +- com.github.ben-manes.caffeine:caffeine:jar:2.8.6:compile
[INFO] |  |  +- org.checkerframework:checker-qual:jar:3.7.0:compile
[INFO] |  |  \- com.google.errorprone:error_prone_annotations:jar:2.4.0:compile
[INFO] |  \- org.springframework:spring-context-support:jar:5.2.10.RELEASE:compile
[INFO] |     +- org.springframework:spring-beans:jar:5.2.10.RELEASE:compile
[INFO] |     \- org.springframework:spring-context:jar:5.2.10.RELEASE:compile
[INFO] +- io.micrometer:micrometer-core:jar:1.5.6:compile
[INFO] |  +- org.hdrhistogram:HdrHistogram:jar:2.1.12:compile
[INFO] |  \- org.latencyutils:LatencyUtils:jar:2.0.3:runtime
[INFO] +- org.springframework.boot:spring-boot-actuator-autoconfigure:jar:2.3.5.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot-actuator:jar:2.3.5.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot:jar:2.3.5.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot-autoconfigure:jar:2.3.5.RELEASE:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.11.3:compile
[INFO] |  |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.11.3:compile
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-core:jar:2.11.3:compile
[INFO] |  \- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:jar:2.11.3:compile
[INFO] +- org.springframework.boot:spring-boot-starter-web:jar:2.3.5.RELEASE:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter:jar:2.3.5.RELEASE:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-starter-logging:jar:2.3.5.RELEASE:compile
[INFO] |  |  |  +- ch.qos.logback:logback-classic:jar:1.2.3:compile
[INFO] |  |  |  |  \- ch.qos.logback:logback-core:jar:1.2.3:compile
[INFO] |  |  |  +- org.apache.logging.log4j:log4j-to-slf4j:jar:2.13.3:compile
[INFO] |  |  |  |  \- org.apache.logging.log4j:log4j-api:jar:2.13.3:compile
[INFO] |  |  |  \- org.slf4j:jul-to-slf4j:jar:1.7.30:compile
[INFO] |  |  +- jakarta.annotation:jakarta.annotation-api:jar:1.3.5:compile
[INFO] |  |  \- org.yaml:snakeyaml:jar:1.26:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-json:jar:2.3.5.RELEASE:compile
[INFO] |  |  +- com.fasterxml.jackson.datatype:jackson-datatype-jdk8:jar:2.11.3:compile
[INFO] |  |  \- com.fasterxml.jackson.module:jackson-module-parameter-names:jar:2.11.3:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-tomcat:jar:2.3.5.RELEASE:compile
[INFO] |  |  +- org.apache.tomcat.embed:tomcat-embed-core:jar:9.0.39:compile
[INFO] |  |  +- org.glassfish:jakarta.el:jar:3.0.3:compile
[INFO] |  |  \- org.apache.tomcat.embed:tomcat-embed-websocket:jar:9.0.39:compile
[INFO] |  +- org.springframework:spring-web:jar:5.2.10.RELEASE:compile
[INFO] |  \- org.springframework:spring-webmvc:jar:5.2.10.RELEASE:compile
[INFO] |     +- org.springframework:spring-aop:jar:5.2.10.RELEASE:compile
[INFO] |     \- org.springframework:spring-expression:jar:5.2.10.RELEASE:compile
[INFO] +- org.springframework.boot:spring-boot-starter-test:jar:2.3.5.RELEASE:test
[INFO] |  +- org.springframework.boot:spring-boot-test:jar:2.3.5.RELEASE:test
[INFO] |  +- org.springframework.boot:spring-boot-test-autoconfigure:jar:2.3.5.RELEASE:test
[INFO] |  +- com.jayway.jsonpath:json-path:jar:2.4.0:test
[INFO] |  |  +- net.minidev:json-smart:jar:2.3:test
[INFO] |  |  |  \- net.minidev:accessors-smart:jar:1.2:test
[INFO] |  |  |     \- org.ow2.asm:asm:jar:5.0.4:test
[INFO] |  |  \- org.slf4j:slf4j-api:jar:1.7.30:compile
[INFO] |  +- jakarta.xml.bind:jakarta.xml.bind-api:jar:2.3.3:test
[INFO] |  |  \- jakarta.activation:jakarta.activation-api:jar:1.2.2:test
[INFO] |  +- org.assertj:assertj-core:jar:3.16.1:test
[INFO] |  +- org.hamcrest:hamcrest:jar:2.2:test
[INFO] |  +- org.junit.jupiter:junit-jupiter:jar:5.7.1:test
[INFO] |  |  +- org.junit.jupiter:junit-jupiter-api:jar:5.7.1:test
[INFO] |  |  |  +- org.opentest4j:opentest4j:jar:1.2.0:test
[INFO] |  |  |  \- org.junit.platform:junit-platform-commons:jar:1.6.3:test
[INFO] |  |  +- org.junit.jupiter:junit-jupiter-params:jar:5.6.3:test
[INFO] |  |  \- org.junit.jupiter:junit-jupiter-engine:jar:5.7.1:test
[INFO] |  |     \- org.junit.platform:junit-platform-engine:jar:1.7.1:test
[INFO] |  +- org.mockito:mockito-core:jar:3.3.3:test
[INFO] |  |  +- net.bytebuddy:byte-buddy:jar:1.10.17:test
[INFO] |  |  +- net.bytebuddy:byte-buddy-agent:jar:1.10.17:test
[INFO] |  |  \- org.objenesis:objenesis:jar:2.6:test
[INFO] |  +- org.mockito:mockito-junit-jupiter:jar:3.3.3:test
[INFO] |  +- org.skyscreamer:jsonassert:jar:1.5.0:test
[INFO] |  |  \- com.vaadin.external.google:android-json:jar:0.0.20131108.vaadin1:test
[INFO] |  +- org.springframework:spring-core:jar:5.2.10.RELEASE:compile
[INFO] |  |  \- org.springframework:spring-jcl:jar:5.2.10.RELEASE:compile
[INFO] |  +- org.springframework:spring-test:jar:5.2.10.RELEASE:test
[INFO] |  \- org.xmlunit:xmlunit-core:jar:2.7.0:test
[INFO] \- io.cucumber:cucumber-core:jar:6.10.1-SNAPSHOT:test
[INFO]    +- io.cucumber:cucumber-gherkin:jar:6.10.1-SNAPSHOT:test
[INFO]    +- io.cucumber:cucumber-gherkin-messages:jar:6.10.1-SNAPSHOT:test
[INFO]    +- io.cucumber:messages:jar:14.0.1:test
[INFO]    +- io.cucumber:tag-expressions:jar:3.0.0:test
[INFO]    +- io.cucumber:cucumber-expressions:jar:10.3.0:test
[INFO]    +- io.cucumber:datatable:jar:3.5.0:test
[INFO]    +- io.cucumber:cucumber-plugin:jar:6.10.1-SNAPSHOT:test
[INFO]    +- io.cucumber:docstring:jar:6.10.1-SNAPSHOT:test
[INFO]    +- io.cucumber:html-formatter:jar:12.0.0:test
[INFO]    +- io.cucumber:create-meta:jar:3.0.0:test
[INFO]    \- org.apiguardian:apiguardian-api:jar:1.1.1:test
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.627 s
[INFO] Finished at: 2021-03-06T20:06:43Z
[INFO] ------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by stepio** [comment]
I don't like this part: `<version>${project.version}</version>`. But it works - added the test description.

As a plan B I considered replacing all the properties with the relevant values, but it seems that it's unnecessary.

File: bom/pom.xml [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    <artifactId>cucumber-bom</artifactId>
+    <packaging>pom</packaging>
+    <name>Cucumber-JVM: Bill of Materials</name>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>io.cucumber</groupId>
+                <artifactId>cucumber-core</artifactId>
+                <version>${project.version}</version>
```
Comment (mpkorstanje) [inline bom/pom.xml:22]: What happens when use this project as a parent pom with a different project version?

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
Looks pretty good, but we now have a duplication between the root `pom.xml` and the `bom/pom.xml`. Can this be cleaned up somehow? Either by reusing the bom in all submodules or including it in the root `pom.xml`?

File: bom/pom.xml (hunk 1 above)
Comment (stepio) [reply bom/pom.xml:22]: Getting an error:
```
[ERROR] Failed to execute goal on project coffee-boots-example: Could not resolve dependencies for project io.github.stepio.examples:coffee-boots-example:jar:0.0.1-SNAPSHOT: Could not find artifact io.cucumber:cucumber-core:jar:0.0.1-SNAPSHOT -> [Help 1]
```
Thanks, that's the case I was looking for, will play more with plan B.

**Comment by stepio** [comment]
> reusing the bom in all submodules

This option means cycle dependency which is detected by Maven if I keep bom as a submodule:
```
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] The dependencies of type=pom and with scope=import form a cycle: io.cucumber:cucumber-bom:6.10.1-SNAPSHOT -> io.cucumber:cucumber-bom:6.10.1-SNAPSHOT @ io.cucumber:cucumber-bom:6.10.1-SNAPSHOT
```

And I don't think that moving bom into a separate repository would solve that either.

> including it in the root pom.xml

To be honest, one can say that there is already a BOM in the project, cause this also works:

```xml
<dependency>
    <groupId>io.cucumber</groupId>
    <artifactId>cucumber-jvm</artifactId>
    <version>6.10.1-SNAPSHOT</version>
    <type>pom</type>
    <scope>import</scope>
</dependency>
```

So the only challenge in this case would be to replace `${project.version}` with the correct value at build (release) time.

**Comment by mpkorstanje** [comment]
I don't see how reusing the bom in all submodules (except the bom) would cause a dependency cycle. Did you add it the root module rather then the sub modules?

**Comment by mpkorstanje** [comment]
Superseded by #2363.

**Pull request closed without merging**